### PR TITLE
Module refactor

### PR DESCRIPTION
**Pull request by fatmcgav**
Raising PR early to get feedback on changes etc. 

Summary of changes:
- Rename resources to `pure_*` to avoid clashing with core Puppet resources.
- Support specifying a REST API version as part of the `device_url`
- Add prefetch support to resources, which removes the need to query the Pure API on every resource, instead getting a list in one call.
- Add support for `wwnlist` on `pure_host`, and make `wwnlist` and `iqnlist` values updateable.
- Add support to `pure_volume` resource for updating the volume size.
- Add support to `pure_connection` for composite namevars.
- Add unit tests to types and providers. 

TODO: 

- [ ] Docs
- [ ] Squash commits

### CODE REVIEW

**Comment by sdodsley** [comment]
@fatmcgav Thanks for this PR. I'll get one of our developers to have a look in the next couple of days and give some feedback.

**Comment by sdodsley** [comment]
@fatmcgav could you summarize this pull request?
a. Requirement / Problem being solved
b. Any notes on the architecture of the patch (e.g. created new classes, objects etc).

**Comment by fatmcgav** [comment]
@sdodsley Sure, I've added a summary of changes to the description.

**Comment by foone** [comment]
Looks good! I see no issues, and this looks like makes some good refactoringings name-wise, and more unit tests are always good.

+1 to merging.

**Review by sdodsley — APPROVED** [review]

**Comment by sdodsley** [comment]
@fatmcgav Is this now complete? If so can you remove the [WIP] flag

**Comment by fatmcgav** [comment]
@sdodsley I'm happy for this to be merged in.
